docs/developer/api-versioning.md

```diff
@@ -0,0 +1,391 @@
+worse with federation taking off: if we add a new end-point to the
+API, the backend will still be able to handle older clients that
+simply don't know about that end-point, but the client won't handle
+old backends well, since it will try to call the end-point, and fail.
+
+The new approach outlined here therefore supports API versions.  Every
+API version is only compatible with itself, but every node in the
+network can support a *set of API versions*.  A HTTP client can query
+the set of supported versions from an HTTP server, and then pick one
+that works for it.
+
+In the following, we will refer to HTTP clients as "clients", no
+matter whether it is a backend talking to another backend
+(federation); and to HTTP servers as "server", no matter which API is
+it serving (federation or app).  This approach is intended to work for
+both.
+
+
+# Versions and servant routing tables
+
+All routing tables for which a new version is born will be changed
+into taking the version number as a parameter, which is added as a
+prefix to every route:
+
+```haskell
+data Api (version :: Symbol) routes = Api
+  { getUnqualifiedConversation ::
+      routes
+        :- version
+        :> Summary "Get a conversation by ID"
+        :> ZLocalUser
+        :> "conversations"
+        :> Capture "cnv" ConvId
+        :> Get '[Servant.JSON] Conversation,
+    getConversation ::
+      routes
+        :- version
+        :> Summary "Get a conversation by ID"
+        :> ZLocalUser
+        :> "conversations"
+        :> QualifiedCapture "cnv" ConvId
+        :> Get '[Servant.JSON] Conversation,
+    [...]
+  }
+```
+
+APIs of all the supported versions can be composed like this:
+
+```haskell
+type ServantAPI =
+  ToServantApi (Api "v1")
+    :<|> ToServantApi (Api "v2")
+    :<|> ToServantApi (Api "v4") -- v3 is broken
+```
+
+
+## Changes between versions
+
+The point of having versions is of course not that all of them look
+exactly the same except for their prefix.  The point is that there are
+other things that change between versions.
+
+There are essentially two categories of changes:
+
+1. **data**: request or response bodies, variable path segments,
+   possible headers or status codes (as in `UVerb` or `MultiVerb`),
+   etc.
+2. **structure**: literal path segments, verb, the version itself, ...
+
+
+## Changes in the data
+
+If a data type in request, response, variable path segments, or anywhere else
+changes, introduce a type family parameterized in the version.
+
+```haskell
+[...]
+    getConversation ::
+      routes
+        :- version
+        :> Summary "Get a conversation by ID"
+        :> ZLocalUser
+        :> "conversations"
+        :> QualifiedCapture "cnv" ConvId
+        :> Get '[Servant.JSON] (ConversationV version),
+[...]
+
+type family ConversationV (version :: Symbol) :: * where
+  ConversationV "v1" = Conversation
+  ConversationV "v2" = Conversation
+  ConversationV "v4" = ConversationV4
```

Comment:
I'm expecting that we'll never support more than 10-15 versions, so I don't think the boilerplate will be too painful (unless I'm missing something).

I'm open to looking into more elegant solutions, but I would probably favour simplicity over brevity.

Reply:
The problem is not as much about being concise, but the amount of changes required to bump version. I'm assuming bumping version is going to be routine, so it would be pretty annoying if we have to go around the whole of wire-api and change the type family for every single one of the types there every time we want to bump version.

I would like a system where bumping versions requires only to change the things that are actually affected. But maybe it's too hard to have this property?

Reply:
Ok, I understand now.  Open question!  e6a937390177ce57623c31360d48acb2ec047dd3